### PR TITLE
perf(netlify, netlify-edge): exclude static paths from server handler

### PR DESCRIPTION
**Pull request by serhalp**
### 🔗 Linked issue

N/A - discussed on Discord

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme, or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

The netlify preset was already using `preferStatic: true` which lets existing static assets take precedence over the function, but this now also avoids a function invocation for static paths that don't exist, i.e. this avoids an unnecessary dynamic 404 that could be served directly from the CDN.

The `netlify-edge` preset wasn't excluding anything, so this addresses both the 404 case and the existing asset case.

The 404 case is important because browsers frequently attempt to request hashed assets from previous deploys that have been invalidated. There's no reason for this to go through functions, since we know that the whole parent path is static.

This PR also adds tests for the `netlify-edge` preset, which were missing. They're mostly copied from the `netlify` preset tests.

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/presets/netlify/utils.ts [hunk 1]
```diff
@@ -92,6 +92,13 @@ export async function writeHeaders(nitro: Nitro) {
   await fsp.writeFile(headersPath, contents);
 }
 
+export function getStaticPaths(nitro: Nitro): string[] {
+  const publicAssets = nitro.options.publicAssets.filter(
+    (dir) => dir.fallthrough !== true && dir.baseURL && dir.baseURL !== "/"
```
Comment (serhalp) [inline src/presets/netlify/utils.ts:97]: @pi0 is this `dir.baseUrl !== "/"` case right? I was seeing this be included when running the tests and I saw similar exclusions in some other presets. I'm not sure I understand what this is doing though.

File: src/presets/netlify/utils.ts [hunk 2]
```diff
@@ -92,6 +92,13 @@ export async function writeHeaders(nitro: Nitro) {
...
+  );
+  return ["/.netlify/*", ...publicAssets.map((dir) => `${dir.baseURL}/*`)];
```
Comment (serhalp) [inline src/presets/netlify/utils.ts:99]: @pi0 are these `baseURL`s guaranteed to be prefixed with a leading slash? I see some presets handling that but I'm not sure if it's necessary. (This Netlify config does require it.)

File: test/presets/netlify.test.ts [hunk 3]
```diff
@@ -86,6 +118,24 @@ describe("nitro:preset:netlify", async () => {
...
+          "utf8"
+        );
+        expect(serverFunctionFile).toEqual(
+          `
+export { default } from "./main.mjs";
+export const config = {
+  name: "server handler",
+  generator: "mock-framework@1.2.3",
+  path: "/*",
+  excludedPath: ["/.netlify/*","/icons/*","/nested/fonts/*","/build/*"],
```
Comment (serhalp) [inline test/presets/netlify.test.ts:133]: @pi0 where is `/build/*` coming from here? I couldn't quite trace it. I'd like to configure it explicitly in the test setup if possible.

File: test/presets/netlify.test.ts [hunk 4]
```diff
@@ -137,3 +187,139 @@ describe("nitro:preset:netlify", async () => {
...
+        const res = await handler(req, {} as EdgeFunctionContext);
+        if (!(res instanceof Response))
+          // The Netlify Edge Function handler API allows returning `undefined` but this
+          // test helper only supports a Response or this shape.
+          return {
+            data: undefined,
+            status: 404,
+            headers: {},
+          };
+        return res;
```
Comment (serhalp) [inline test/presets/netlify.test.ts:258]: @pi0 I've got one test failure from the `testNitro()` cases that I don't understand:

```
 FAIL  test/presets/netlify.test.ts > nitro:preset:netlify-edge > API Works
AssertionError: expected undefined to match object { message: 'Hello API' }

- Expected:
Object {
  "message": "Hello API",
}

+ Received:
undefined

 ❯ test/tests.ts:229:26
    227|   it("API Works", async () => {
    228|     const { data: helloData } = await callHandler({ url: "/api/hello" });
    229|     expect(helloData).to.toMatchObject({ message: "Hello API" });
       |                          ^
    230|
    231|     if (ctx.nitro?.options.serveStatic) {
```

It's hitting this case where the edge function handler is returning `undefined`. Any tips? I'm not sure what's being tested here and why it would be behaving differently for netlify-edge.

File: src/presets/netlify/utils.ts (hunk 1 above)
Comment (pi0) [reply src/presets/netlify/utils.ts:97]: It is correct. The check is to make sure we don't include root `/` patterns (that imply fallthrough behavior) to exclude. Nitro usually infers `fallthrough` properly but it might not be set correctly (ie, after creating nitro instance via wrapper frameworks)

File: test/presets/netlify.test.ts (hunk 4 above)
Comment (pi0) [reply test/presets/netlify.test.ts:258]: `/api/hello` is expected to be prerendered (an static asset in output public dir)

File: test/presets/netlify.test.ts [hunk 5]
```diff
@@ -1,14 +1,44 @@
...
+import type { Context as EdgeFunctionContext } from "@netlify/edge-functions";
 import { resolve } from "pathe";
 import { describe, expect, it } from "vitest";
 import { getPresetTmpDir, setupTest, testNitro } from "../tests";
 
 describe("nitro:preset:netlify", async () => {
+  const publicDir = resolve(getPresetTmpDir("netlify"), "dist");
   const ctx = await setupTest("netlify", {
     config: {
+      framework: {
```
Comment (pi0) [inline test/presets/netlify.test.ts:12]: I think we can directly add `framework` to the main fixture nitro.config file

File: test/presets/netlify.test.ts (hunk 3 above)
Comment (pi0) [reply test/presets/netlify.test.ts:133]: test fixture i suppose

File: test/presets/netlify.test.ts [hunk 6]
```diff
@@ -137,3 +187,139 @@ describe("nitro:preset:netlify", async () => {
...
+
+describe("nitro:preset:netlify-edge", async () => {
+  const publicDir = resolve(getPresetTmpDir("netlify-edge"), "dist");
+  const ctx = await setupTest("netlify-edge", {
+    config: {
+      framework: {
+        name: "mock-framework",
+        version: "1.2.3",
+      },
+      publicAssets: [
```
Comment (pi0) [inline test/presets/netlify.test.ts:199]: This seems like duplicate logic we can at least share between two netlify/edge tests.

Ideally, I would love to extend the main fixture with more cases of public assets so that we increase overall coverage instead of per-preset tests.

**Comment by pi0** [comment]
Thanks for the PR dear @serhalp 

It seems main blockers on PR are addition of tests (which are so nice) -- although i don't want the main improvement to be missed for 2.10 release, perhaps we can extract work on tests to an additional PR? 

Main addition seems could be tested easily both manually or with a simple unit test for `getStaticPaths` utility.

File: test/tests.ts [hunk 7]
```diff
@@ -223,16 +224,22 @@ export function testNitro(
     _handler = await getHandler();
   }, 25_000);
 
-  it("API Works", async () => {
-    const { data: helloData } = await callHandler({ url: "/api/hello" });
-    expect(helloData).to.toMatchObject({ message: "Hello API" });
+  // netlify-edge intentionally ignores prerendered routes
+  it.skipIf(ctx.preset === "netlify-edge")(
```
Comment (serhalp) [inline test/tests.ts:228]: https://github.com/serhalp/nitro/blob/ce72ea3fd1ae59cf722a9868d8e9cc99eb6b0aa1/src/presets/netlify/runtime/netlify-edge.ts#L12-L14

happy to handle this a different way in a follow-up PR if you'd like

File: src/presets/netlify/utils.ts (hunk 2 above)
Comment (pi0) [reply src/presets/netlify/utils.ts:99]: it should be normalized but we can use `joinURL` from `ufo` to make sure of it.

File: src/presets/netlify/utils.ts [hunk 8]
```diff
@@ -92,6 +92,13 @@ export async function writeHeaders(nitro: Nitro) {
   await fsp.writeFile(headersPath, contents);
 }
 
+export function getStaticPaths(nitro: Nitro): string[] {
```
Comment (pi0) [inline src/presets/netlify/utils.ts:95]: This utility can accept `publicAssets` directly. This way we can write simple unit tests for it without all new complexity being added (both fixture and tests and be simplified)

File: test/tests.ts (hunk 7 above)
Comment (pi0) [reply test/tests.ts:228]: I would suggest to use unit test for new functionality and avoid any changes here for now.

Comment (serhalp) [reply test/tests.ts:228]: so you'd prefer to split out the addition of the missing tests for netlify-edge into a separate PR?

Comment (pi0) [reply test/tests.ts:228]: Sure

**Comment by serhalp** [comment]
@pi0 I made all the requested changes.

Here's a manual QA against a nuxt site:
- site homepage is rendered via SSR: https://672396fce8c6141da780bc41--nuxt-test-static-assets-fix.netlify.app
- a nonexistent path under `/_nuxt` returns a 404 directly from the CDN without invoking a function: https://672396fce8c6141da780bc41--nuxt-test-static-assets-fix.netlify.app/_nuxt/fake-old-file-123.css
- a real path under `/_nuxt` returns a 200 directly from the CDN without invoking a function: https://672396fce8c6141da780bc41--nuxt-test-static-assets-fix.netlify.app/_nuxt/entry.CQDx9R-V.css

**Comment by serhalp** [comment]
And for good measure here's one against `netlify-edge`:
- site homepage is rendered via SSR: https://67239a3eeb45272890169516--nuxt-test-static-assets-fix.netlify.app
- a nonexistent path under `/_nuxt` returns a 404 directly from the CDN without invoking a function: https://67239a3eeb45272890169516--nuxt-test-static-assets-fix.netlify.app/_nuxt/fake-old-file-123.css
- a real path under `/_nuxt` returns a 200 directly from the CDN without invoking a function: https://67239a3eeb45272890169516--nuxt-test-static-assets-fix.netlify.app/_nuxt/entry.CQDx9R-V.css

File: test/tests.ts [hunk 9]
```diff
@@ -224,22 +224,16 @@ export function testNitro(
     _handler = await getHandler();
   }, 25_000);
 
-  // netlify-edge intentionally ignores prerendered routes
```
Comment (serhalp) [inline test/tests.ts]: oops, forgot to undo this one, my bad! thanks for cleaning up

**Review by pi0 — APPROVED** [review]

**Comment by pi0** [comment]
Thanks ❤️
